docs/design/generic.md

```diff
@@ -0,0 +1,95 @@
+# Generic artifact fetching
+
+## Introduction
+
+This document will describe high-level implementation overview for supporting generic artifact fetching in cachi2.
+Up until now cachi2 has only supported package managers for various ecosystems and languages.
+However, there are a couple of use-cases where language non-specific artifacts need to be pre-fetched in order to satisfy
+requirements of a hermetic build.
+
+## Context
+
+For context, generic artifact fetching is a use-case of its own (e.g. [OVAL feeds](https://github.com/CISecurity/OVALRepo),
+AI models), it is also necessary precursor for implementing support for fetching maven artifacts, which won't be covered
+in this design, but in a followup document.
+
+## Design
+
+In this section, I will try to cover individual parts of the design.
+
+### Source repository
+
+This section will describe the structure of the source repository, that will serve as an input to cachi2. The idea is to
+define a cachi2 lockfile that will specify individual artifacts to fetch along with necessary metadata - e.g. checksums.
+The format chosen for this lockfile is yaml, and will include [purl](https://github.com/package-url/purl-spec) for each
+of the fetched artifacts. This decision was made mainly because it allows for followup implementation of maven support,
+with accurate SBOM information. Here's an example of such a lockfile.
+
+```yaml
+artifacts:
+  - purl: pkg:generic/granite-model?download_url=https://huggingface.co/instructlab/granite-7b-lab/resolve/main/model-00001-of-00003.safetensors?download=true
+    target: granite-model.safetensors
+    checksums:
+      sha256: 07123e1f482356c415f684407a3b8723e10b2cbbc0b8fcd6282c49d37c9c1abc
+```
+
+#### Lockfile format and validation
+
+##### purl (required)
+
+At this point, the only purl type allowed would be `pkg:generic`. This is because cachi2 has no good way of verifying
+additional properties of the fetched artifact that could be included in the resulting SBOM. This should create a strong
+incentive to use this feature in the only truly necessary cases, because it will generate low-quality SBOM components,
+as compared to using other package managers provided by cachi2. Additionally, the only allowed qualifier should be `download_url`.
+
+#### target (optional)
```

Comment:
I'm still not sure about this feature. I know the push here is to avoid breaking the workflow of some users that will migrate to Cachi2, but I think we should primarily see this from a more generic perspective. Will it benefit users of the generic fetching? Maybe it's useful to allow users to specify a filename that differs from what exists in the remote server they'll download the file from. 

In any case, to follow the precedent we have, I'd say that the `target` must be a subpath of `cachi2-output/deps/generic`

Reply:
You're right. I was trying to find a justification for this feature looking from a more generic perspective. But as long as it falls under `cachi2-output/deps/generic`, I'm ok with this feature.

Reply:
In any case, we might need to do some check to avoid naming collisions for generic artifacts downloads.

Reply:
> In any case, we might need to do some check to avoid naming collisions for generic artifacts downloads.
Agreed, I have included that in this section. Closing this as resolved, feel free to unresolve if you feel there's more to discuss.